### PR TITLE
(CDAP-16324) Support macro for Joiner, Dedup and Distinct transform fields

### PR DESCRIPTION
**Pull request by rmstar**

### CODE REVIEW

File: core-plugins/src/main/java/io/cdap/plugin/batch/joiner/JoinerConfig.java [hunk 1]
```diff
@@ -177,9 +189,18 @@ private boolean isAliasPresent(Iterable<String> stageOldNameAliasPair) {
...
   Set<String> getInputs() {
-    if (!Strings.isNullOrEmpty(requiredInputs)) {
+    if (!Strings.isNullOrEmpty(requiredInputs) && !containsMacro(JoinerConfig.REQUIRED_INPUTS)) {
       return ImmutableSet.copyOf(Splitter.on(',').trimResults().omitEmptyStrings().split(requiredInputs));
     }
     return ImmutableSet.of();
   }
+
+  void validateJoinKeySchemas(Map<String, Schema> inputSchemas, Map<String, List<String>> joinKeys,
+        FailureCollector collector) {
```
Comment (CuriousVini) [inline core-plugins/src/main/java/io/cdap/plugin/batch/joiner/JoinerConfig.java]: nit: indentation

File: core-plugins/src/main/java/io/cdap/plugin/batch/joiner/JoinerConfig.java [hunk 2]
```diff
@@ -177,9 +189,18 @@ private boolean isAliasPresent(Iterable<String> stageOldNameAliasPair) {
...
+      if (!containsMacro(JoinerConfig.JOIN_KEYS) && joinKeys.size() != inputSchemas.size()) {
+        collector.addFailure("There should be join keys present from each stage.",
+            "Ensure join keys are present from each stage.")
+            .withConfigProperty(JoinerConfig.JOIN_KEYS);
+      }
```
Comment (CuriousVini) [inline core-plugins/src/main/java/io/cdap/plugin/batch/joiner/JoinerConfig.java]: nit: indentation
Please import the settings jar and format the code.

File: core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java [hunk 3]
```diff
@@ -170,11 +174,11 @@ public void prepareRun(BatchJoinerContext context) {
   @Override
   public void initialize(BatchJoinerRuntimeContext context) {
     FailureCollector collector = context.getFailureCollector();
-    init(context.getInputSchemas(), collector);
-    collector.getOrThrowException();
-
     Map<String, Schema> inputSchemas = context.getInputSchemas();
-    outputSchema = context.getOutputSchema();
+    init(inputSchemas, collector);
+    collector.getOrThrowException();
```
Comment (CuriousVini) [inline core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java:179]: do we need this?

File: core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java [hunk 4]
```diff
@@ -82,7 +82,11 @@ public void configurePipeline(MultiInputPipelineConfigurer pipelineConfigurer) {
     init(inputSchemas, collector);
     collector.getOrThrowException();
     //validate the input schema and get the output schema for it
-    stageConfigurer.setOutputSchema(getOutputSchema(inputSchemas, collector));
+    Schema outputSchema = getOutputSchema(inputSchemas, collector);
```
Comment (CuriousVini) [inline core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java:85]: does getOutputSchema handle null join keys and other configs?

File: core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java [hunk 5]
```diff
@@ -279,7 +278,12 @@ Schema getOutputSchema(Map<String, Schema> inputSchemas, FailureCollector collec
     requiredInputs = conf.getInputs();
     perStageSelectedFields = conf.getPerStageSelectedFields();
     duplicateFields = ArrayListMultimap.create();
-    return Schema.recordOf("join.output", getOutputFields(createOutputFieldInfos(inputSchemas, collector)));
+    List<Schema.Field> outputFields = getOutputFields(createOutputFieldInfos(inputSchemas, collector));
+    if (outputFields.isEmpty()) {
+      return Schema.recordOf("join.output");
```
Comment (CuriousVini) [inline core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java]: shouldnt this return null?

**Review by CuriousVini — COMMENTED** [review]
a few comments.
The jira also mentions dedup plugin. Assuming there will be another PR for it?

File: core-plugins/src/main/java/io/cdap/plugin/batch/joiner/JoinerConfig.java (hunk 1 above)
Comment (rmstar) [reply core-plugins/src/main/java/io/cdap/plugin/batch/joiner/JoinerConfig.java]: fixed.

File: core-plugins/src/main/java/io/cdap/plugin/batch/joiner/JoinerConfig.java (hunk 2 above)
Comment (rmstar) [reply core-plugins/src/main/java/io/cdap/plugin/batch/joiner/JoinerConfig.java]: fixed.

File: core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java (hunk 5 above)
Comment (rmstar) [reply core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java]: Fixed to return null.

File: core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java (hunk 4 above)
Comment (rmstar) [reply core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java:85]: Yes.
null join keys (configured by macro) is handled by JoinerConfig.validateJoinKeySchemas().
JoinerConfig.getInputs() returns empty set when requiredInputs is null.
getOutputSchema returns null if selectedFields is null.

File: core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java (hunk 3 above)
Comment (rmstar) [reply core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java:179]: It's needed to handle validation failures in join key validation failures.

**Comment by rmstar** [comment]
I'll open a separate PR for dedupe plugin changes.
Edit: It looks like adding macro support for dedupe and distinct plugins is trivial - just requires a `@Macro` annotation. So I'll update them as part of this PR.

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupConfig.java [hunk 6]
```diff
@@ -52,6 +54,7 @@
     "of field and function is allowed. If this property is not set, one random record will be chosen from the " +
     "group of 'duplicate' records.")
   @Nullable
+  @Macro
```
Comment (CuriousVini) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupConfig.java:57]: are nulls handled for both the macro properties?

**Review by CuriousVini — APPROVED** [review]
just one question, code changes lgtm

**Comment by rmstar** [comment]
> just one question, code changes lgtm

Yes, nulls are already handled for dedupe and distinct. If fields are not specified, all fields are used for dedupe and distinct plugins.

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupConfig.java (hunk 6 above)
Comment (rmstar) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupConfig.java:57]: Yes

**Pull request merged**